components/scream/src/share/io/scream_io_utils.hpp

```diff
@@ -44,12 +45,54 @@ inline OutputAvgType str2avg (const std::string& s) {
 struct IOControl {
   // A non-positive frequency can be used to signal IO disabled
   int frequency = -1;
-  int nsteps_since_last_write;
+  int nsamples_since_last_write;  // Needed when updating output data, such as with the OAT::Average flag
+  util::TimeStamp timestamp_of_last_write;
   std::string frequency_units;
 
-  bool is_write_step () {
-    return frequency>0 && (nsteps_since_last_write % frequency == 0);
-  }
+  bool is_write_step (const util::TimeStamp& ts) {
+    // Mini-routine to determine if it is time to write output to file.
+    // The current allowable options are nsteps, nsecs, nmins, nhours, ndays, nmonths, nyears
+    // We query the frequency_units string value to determine which option it is.
+    bool ret = false;
+    if (frequency > 0 && frequency_units != "never") {
+      auto ts_diff = (ts-timestamp_of_last_write);
+      if (frequency_units == "nsteps") {
+        // Just use the num_steps from timestamps
+        return ((ts.get_num_steps()-timestamp_of_last_write.get_num_steps()) % frequency == 0);
+      // We will need to use timestamp information
+      } else if (frequency_units == "nsecs") {
+        ret = ((ts_diff > 0) && (ts_diff % frequency == 0));
+      } else if (frequency_units == "nmins") {
+        ret = (ts_diff >= 60) && (ts_diff % (frequency*60) == 0);
+      } else if (frequency_units == "nhours") {
+        ret = (ts_diff >= 3600) && (ts_diff % (frequency*3600) == 0);
+      } else if (frequency_units == "ndays") {
+        ret = (ts_diff >= 86400) && (ts_diff % (frequency*86400) == 0);
+      } else if (frequency_units == "nmonths" || frequency_units == "nyears") {
+        // For months and years we need to be careful, can't just divide ts_diff by a set value.
+        // First we make sure that if we are the same day of the month and at the same time of day.
+        if (ts.get_day() == timestamp_of_last_write.get_day() &&
```

Comment:
This line here is _assuming_ that timestepping will end up on the same time of day every day. Is it safe to assume that atm timesteps will _always_ be such that there's an integer number of steps in a day? I think the answer is yes, due to ATM_NCPL cime entry, but wanted to double check.

Reply:
Uhm, you also check secofday, so it should pick the right time of day. I'm ok without adding a check. You don't have dt in this routine after all, so you can't do it here.